### PR TITLE
refactor(revision): subject revision

### PR DESCRIPTION
**Pull request by Sociosarbis**
rewrite subject revision apis in go

issue: #121

### CODE REVIEW

**Comment by Trim21** [comment]
能不能一起加个添加历史版本的方法





------------------ Original ------------------
From: notifications ***@***.***&gt;
Date: Wed,Mar 23,2022 6:32 PM
To: bangumi/server ***@***.***&gt;
Cc: Subscribed ***@***.***&gt;
Subject: Re: [bangumi/server] refactor(revision): subject revision (PR #133)




rewrite subject revision apis in go
 
issue: #121
 
 
You can view, comment on, or merge this pull request online at:
 
&nbsp; https://github.com/bangumi/server/pull/133
 
Commit Summary
    
2f50757  refactor(revision): subject revision 
 
File Changes 
 
(10 files)
    
     M     domain/revision.go     (8)      

     M     mocks/RevisionRepo.go     (139)      

     M     model/revision.go     (17)      

     M     revision/mysql_repository.go     (66)      

     M     revision/mysql_repository_test.go     (42)      

     M     web/handler.go     (2)      

     M     web/handler/revision.go     (119)      

     M     web/handler/revision_test.go     (72)      

     M     web/res/revision.go     (22)      

     M     web/spec.yaml     (55)    
 
Patch Links:
    
https://github.com/bangumi/server/pull/133.patch   

https://github.com/bangumi/server/pull/133.diff

File: revision/mysql_repository_test.go [hunk 1]
```diff
@@ -73,3 +76,38 @@ func TestListPersonRelated(t *testing.T) {
...
+	t.Parallel()
+
+	repo := getRepo(t)
+
+	r, err := repo.GetSubjectRelated(context.Background(), 718391)
+	require.NoError(t, err)
+	require.Equal(t, uint32(718391), r.ID)
+	data, ok := r.Data.(*model.SubjectRevisionData)
+	require.True(t, ok)
+	require.Equal(t, data.Name, "第一次的親密接觸")
```
Comment (Trim21) [inline revision/mysql_repository_test.go:91]:
```suggestion
	require.Equal(t, "第一次的親密接觸", data.Name)
```

File: revision/mysql_repository_test.go [hunk 2]
```diff
@@ -47,8 +48,10 @@ func TestGetPersonRelatedBasic(t *testing.T) {
...
 	require.NoError(t, err)
 	require.Equal(t, uint32(348475), r.ID)
-	data := handler.CastPersonData(r.Data)
-	d, ok := data["348475"]
+	data, ok := r.Data.(map[string]interface{})
+	require.True(t, ok)
+	dat := handler.CastPersonData(data)
+	d, ok := dat["348475"]
 	require.True(t, ok)
 	require.Equal(t, d.Name, "森岡浩之")
```
Comment (Trim21) [inline revision/mysql_repository_test.go:56]:
```suggestion
	require.Equal(t, "森岡浩之", d.Name)
```

File: web/handler/revision.go [hunk 3]
```diff
@@ -100,7 +100,80 @@ func (h Handler) GetPersonRevision(c *fiber.Ctx) error {
...
+	for i := range revisions {
+		data[i] = convertModelSubjectRevision(&revisions[i], creatorMap)
+	}
+	response.Data = data
+	return c.JSON(response)
+}
+
+func (h Handler) GetSubjectRevision(c *fiber.Ctx) error {
+	id, err := strparse.Uint32(c.Params("id"))
+	if err != nil || id <= 0 {
```
Comment (Trim21) [inline web/handler/revision.go:163]:
```suggestion
	if err != nil || id == 0 {
```

File: web/handler/revision.go [hunk 4]
```diff
@@ -100,7 +100,80 @@ func (h Handler) GetPersonRevision(c *fiber.Ctx) error {
...
+		return fiber.NewError(fiber.StatusBadRequest, fmt.Sprintf("bad param id: %s", c.Params("id")))
```
Comment (Trim21) [inline web/handler/revision.go:164]:
```suggestion
		return fiber.NewError(fiber.StatusBadRequest, "bad param id: " + strconv.Quote(c.Params("id")))
```

**Comment by Trim21** [comment]
啊，我好像操作错了，这两个review你看到了吗 
image

**Comment by Sociosarbis** [comment]
> 啊，我好像操作错了，这两个review你看到了吗 image

没看到😄。我晚上改改再PR

**Comment by Trim21** [comment]
还有slack里提到的那个问题...

**Pull request merged**